blast: init at 2.10.0

luispedro opened the pull request:
###### Motivation for this change

BLAST is one of the basic bioinformatics packages

###### Things done


- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

luispedro committed:
blast: init at 2.9.0

This is widely used in bioinformatics.

luispedro committed:
blast: init at 2.9.0

FRidh commented:
@GrahamcOfBorg build blast

FRidh commented:
The build seems to fail.

luispedro commented:
My bad, I thought I had built it in a sandbox, but apparently I got fooled by running `nix-build` with `--sandbox`, which does not trigger a rebuild if the package was previously built without `--sandbox`.

Anyway, this package aggressively wants to build using utilities in `/bin` so I had to patch around it.

luispedro commented:
@GrahamcOfBorg build blast

luispedro commented:
I think the issues are fixed now. This is an important package in bioinformatics that'd be great to have in nixpkgs.

FRidh commented:
@GrahamcOfBorg build blast

FRidh commented:
@GrahamcOfBorg build blast

c0bw3b commented on pkgs/applications/science/biology/blast/default.nix:
```diff
@@ -0,0 +1,98 @@
+    patchShebangs $out/bin/update_blastdb.pl
+  '';
+
+
```
One empty line should be enough for readability

luispedro replied:
Will fix & force-push

c0bw3b commented on pkgs/applications/science/biology/blast/default.nix:
```diff
@@ -0,0 +1,98 @@
+  buildInputs = [ coreutils gawk zlib boost bzip2 cpio ];
+  hardeningDisable = [ "format" ];
+
+  patches = [ ./no_slash_bin.patch ];
```
Isn't this going to be a pain to maintain? The whole patch may need to be reworked for each new version

luispedro replied:
Yep. I can replace one or two instances of patching with `substituteInPlace`, but otherwise, this is a pretty hostile build script: it repeatedly tries to call utilities in `/bin/`

luispedro replied:
I replaced the uses of the patch that I could with `substituteInPlace`, but the remaining ones don't seem to lend themselves to doing so.

pschuprikov replied:
Few days before I noticed this PR I was trying to get blast compiling in my `nur-packages`, I used a very coarse-grained patching https://github.com/pschuprikov/nur-packages/blob/d15609f3442ed4dac23d9d782f78cd734221e5b4/pkgs/bioinf/ncbi_blast/generic.nix#L19-L38. 

@luispedro if you feel that could be a better way to go, I can create a PR against your branch to save your time.

luispedro committed:
MIN Reduce patching by using substituteInPlace

orivej commented:
See also #33961. (I'm OK with this new PR, it is just more difficult to maintain.)

luispedro commented:
I wasn't aware of the older PR.

Given the centrality of BLAST in the field and the fact that it's a very stable tool, I think it's best to have something accepted even if it is not always as up to date (for the non-bioinformaticians: it's best to have a 2-year old version of `grep` that none at all).

I personally don't care if it's the older version in #33961 (although maybe that one should have some of the `postInstall` steps here) or this one.

luispedro commented:
Updated against `master` and it still builds/works cleanly.

luispedro committed:
blast: downgrade boost

Newer versions cause errors in the unit_test module

luispedro commented:
Fixed for `master` (in particular, newer `boost` versions cause errors) and rebased.

pschuprikov commented:
I have successfully built it locally. The tests are not being run
```
running tests
no check/test target in Makefile, doing nothing
```
Since you remove the unit tests at `postInstall` due to the lack of external databases, are you sure that`doCheck = true` is any useful? (BTW, `boost` seems to be only needed for tests)

luispedro committed:
Disable tests

Most tests require network use and/or locally-installed databases.

luispedro commented:
You're right. Disabled tests.

markuskowa commented:
@GrahamcOfBorg build blast

markuskowa commented:
The build fails on darwin, Is it supposed to run on a Mac?

pschuprikov commented:
Yes, it is. I'll try to see what's wrong.

pschuprikov commented:
The issue seems to be with the missing frameworks. At least, adding `ApplicationServices` makes it pass the configure step and start building. I'll check the full compilation in the coming days.

pschuprikov commented on pkgs/applications/science/biology/blast/default.nix:
```diff
@@ -0,0 +1,99 @@
+
+  sourceRoot = "ncbi-blast-${version}+-src/c++";
+  
+  configureFlags = [ "--without-makefile-auto-update" ];
```
I think `--with-dll` is very much necessary here, otherwise, `$out/bin` gets some four times as huge.

pschuprikov commented on pkgs/applications/science/biology/blast/default.nix:
```diff
@@ -0,0 +1,99 @@
+  postInstall = ''
+    # BLAST installs some unit test binaries, but several of them cannot be run
+    # without the corresponding test databases, so clean up
+    rm -f $out/bin/*_unit_test
```
These lines should be unnecessary now since without `boost` tests don't get built.

pschuprikov committed:
blast: do not build extra deps, +darwin, +dll

markuskowa commented on pkgs/applications/science/biology/blast/default.nix:
```diff
@@ -0,0 +1,105 @@
+
+  patches = [ ./no_slash_bin.patch ];
+
+  postInstall = ''
```
The `patchShebangs` fixes should not be necessary if you add `perl` also to `buildInputs`. For the shell script it should work out of the box.

luispedro replied:
Thanks

luispedro committed:
MIN Remove unnecessary shebang patch

The bash patch is automatic
The perl patches become automatic if perl is included in buildInputs

suggested by @markuskowa on GH

luispedro commented:
Rebased to master & removed the unnecessary shebang patching.

luispedro committed:
blast : 2.9.0 -> 2.10.0

luispedro commented:
Updated to `2.10.0`. It was a simple version+hash update, so that's a good sign.

markuskowa commented:
`bin/get_species_taxids.sh` fails with:
```
Cannot find Entrez EDirect esearch tool, please see installation in https://www.ncbi.nlm.nih.gov/books/NBK179288/
results/blast/bin/get_species_taxids.sh: line 1: /bin/rm: No such file or directory
```
It looks like there are some hard coded `/bin/...` paths in the script.

luispedro commented:
> Cannot find Entrez EDirect esearch

Ah, yes, that is a tool that must be installed separately. Currently, there is no nix package for it, though.

> It looks like there are some hard coded /bin/... paths in the script.

Aargh, this package really does want to use `/bin/rm` all the time for no good reason.

They also set `$PATH` to include some internal paths (`export PATH=/bin:/usr/bin:/am/ncbiapdata/bin:$HOME/edirect:$PATH).

markuskowa commented:
@luispedro Do you want to fix the `/bin/rm` entries (e.g. with a `substituteInPlace`) so that it works out of the box later or leave it for now?

luispedro committed:
MIN Patch get_species_taxids.sh to not use /bin/rm

luispedro commented:
Fixed the `/bin/rm` dependency, so that, if users have `entrez` installed somehow, it will be useful.

markuskowa commented:
@luispedro there's still a problem with the darwin build.

luispedro commented:
Unfortunately, I have no way of testing on Darwin. I can only restrict the package to Linux if nobody else can work on it.

markuskowa commented:
Restricting the package to linux seems to be the best option for now. If someone else is willing to fix it we can always do it later in a separate PR.

luispedro committed:
MIN Restrict to Linux

luispedro commented:
Restricted to Linux now.

pschuprikov commented:
I do plan to fix the build for Darwin, but it can definitely be done as a separate PR.
